src/dialogs/editor.tsx

```diff
@@ -0,0 +1,128 @@
+import { useDialogs } from 'dialogs';
+import { useInit } from 'hooks';
+import { InlineNotification } from 'cockpit-components-inline-notification';
+
+import "./editor.scss";
+
+const _ = cockpit.gettext;
+
+function useModified<T>(initial: T) {
+    const [modified, setModified] = React.useState<T>(initial);
+
+    React.useEffect(() => {
+        const before_unload = (event: BeforeUnloadEvent) => {
+            event.preventDefault();
+        };
+        if (modified) {
+            window.addEventListener('beforeunload', before_unload);
+            return () => {
+                window.removeEventListener('beforeunload', before_unload);
+            };
+        }
+    }, [modified]);
+
+    return [modified, setModified] as const;
+}
+
+export const EditFileModal = ({ path } : { path: string }) => {
+    const [error, setError] = React.useState<string>();
+    const [content, setContent] = React.useState<string>('');
+    const [modified, setModified] = useModified(false);
+    const Dialogs = useDialogs();
+
+    useInit(async () => {
+        try {
+            setContent(await cockpit.file(path).read());
+        } catch (exc) {
+            setError(cockpit.message(exc as cockpit.JsonObject));
+        }
+    }, [path]);
+
+    const modify = (_ev: unknown, value: string) => {
+        if (content !== value) {
+            setError(undefined);
+            setContent(value);
+            setModified(true);
+        }
+    };
+
+    const save = async () => {
+        if (content !== undefined) {
+            try {
+                await cockpit.file(path).replace(content);
+                setModified(false);
+            } catch (exc) {
+                setError(cockpit.message(exc as cockpit.JsonObject));
+            }
+        }
+    };
+
+    return (
+        <Modal
+          position="top"
+          /* Translators: $0 is a filename */
+          title={(modified ? "⭐ " : "") + cockpit.format(_("Editing $0"), path)}
+          isOpen
+          onClose={Dialogs.close}
+          variant={ModalVariant.large}
+          className='file-editor-modal'
+          footer={
+              <>
+                  <Button
+                    variant="primary"
+                    isDisabled={!modified}
+                    onClick={save}
+                  >
+                      {_("Save")}
+                  </Button>
+                  <Button variant="link" onClick={Dialogs.close}>
+                      {modified ? _("Cancel (discard changes)") : _("Cancel")}
+                  </Button>
+              </>
+          }
+        >
+            <Stack>
+                {error !== undefined &&
+                <InlineNotification
+                  type="danger"
+                  text={error}
+                  isInline
+                />}
+                <TextArea
+                  id='editor-text-area'
+                  className='file-editor'
+                  value={content || ''}
+                  onChange={modify}
```

Comment:
it probably makes sense to also disabled the textarea when there is an error `isDisabled={error !== null}`